idb/postgres/internal/migrations/local_ledger/migration.go

```diff
@@ -0,0 +1,170 @@
+package localledger
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"path"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/algorand/go-algorand-sdk/client/v2/algod"
+	algodConfig "github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/rpcs"
+	"github.com/algorand/indexer/fetcher"
+	"github.com/algorand/indexer/idb"
+	"github.com/algorand/indexer/processor"
+	"github.com/algorand/indexer/processor/blockprocessor"
+	"github.com/algorand/indexer/util"
+	log "github.com/sirupsen/logrus"
+)
+
+// RunMigration executes the migration core functionality.
+func RunMigration(round uint64, opts *idb.IndexerDbOptions) error {
+	logger := log.New()
+	ctx, cf := context.WithCancel(context.Background())
+	defer cf()
+	{
+		cancelCh := make(chan os.Signal, 1)
+		signal.Notify(cancelCh, syscall.SIGTERM, syscall.SIGINT)
+		go func() {
+			<-cancelCh
+			logger.Errorf("Ledger migration interrupted")
+			// exit process if migration is interrupted so that
+			// migration state doesn't get updated in db
+			os.Exit(1)
+		}()
+	}
+
+	var bot fetcher.Fetcher
+	var err error
+	if opts.IndexerDatadir == "" {
+		return fmt.Errorf("RunMigration() err: indexer data directory missing")
+	}
+	// create algod client
+	if opts.AlgodDataDir != "" {
+		bot, err = fetcher.ForDataDir(opts.AlgodDataDir, logger)
+		if err != nil {
+			return fmt.Errorf("RunMigration() err: %w", err)
+		}
+	} else if opts.AlgodAddr != "" && opts.AlgodToken != "" {
+		bot, err = fetcher.ForNetAndToken(opts.AlgodAddr, opts.AlgodToken, logger)
+		if err != nil {
+			return fmt.Errorf("RunMigration() err: %w", err)
+		}
+	} else {
+		return fmt.Errorf("RunMigration() err: unable to create algod client")
+	}
+
+	logger.Info("initializing ledger")
+	genesis, err := getGenesis(bot.Algod())
+	if err != nil {
+		return fmt.Errorf("RunMigration() err: %w", err)
+	}
+	genesisBlock, err := getGenesisBlock(bot.Algod())
+	if err != nil {
+		return fmt.Errorf("RunMigration() err: %w", err)
+	}
+	initState, err := util.CreateInitState(&genesis, &genesisBlock)
+	if err != nil {
+		return fmt.Errorf("RunMigration() err: %w", err)
+	}
+
+	localLedger, err := ledger.OpenLedger(logging.NewLogger(), filepath.Join(path.Dir(opts.IndexerDatadir), "ledger"), false, initState, algodConfig.GetDefaultLocal())
+	if err != nil {
+		return fmt.Errorf("RunMigration() err: %w", err)
+	}
+	defer localLedger.Close()
+	bot.SetNextRound(uint64(localLedger.Latest()) + 1)
+	proc, err := blockprocessor.MakeProcessor(localLedger, nil)
```

Comment:
Sorry for the delayed reviews. These changes are now overlapping, I think this code would be slightly simplified once OpenLedger moves into the constructor.